### DISCUSSION TITLE
Porting OpenSSL to barebones 32 bit RISC-V

### DISCUSSION THREAD

**Discussion by crispy245** [openssl/openssl#23336] [Q&A]
Hi, I been trying to get OpenSSL crosscompiled to RISC-V 32 bits but I seem to get no luck whatsoever.
I got it to compile to 64-bits using `./Configure linux-generic64 --cross-compile-prefix=riscv64-unknown-linux-gnu- --prefix=...` which works however I need to get it compiled to RV32IMAC standard, so I thought just adding the `-march=rv32imac` and `-mabi=ilp32 ` but this constantly gets me errors like 

```
riscv64-unknown-linux-gnu-gcc -fPIC -pthread -march=rv32i -mabi=ilp32 -std=c99 -Wl,-znodelete -shared -Wl,-Bsymbolic  -I/opt/riscv-gnu -Wl,-soname=libcrypto.so.3 \
	-o libcrypto.so.3 -Wl,--version-script=libcrypto.ld \
	crypto/aes/libcrypto-shlib-aes_cbc.o \
	crypto/aes/libcrypto-shlib-aes_cfb.o \
	crypto/aes/libcrypto-shlib-aes_core.o \
	crypto/aes/libcrypto-shlib-aes_ecb.o \
	.
        .
        .
	providers/libcrypto-shlib-defltprov.o \
	providers/libcrypto-shlib-nullprov.o \
	providers/libcrypto-shlib-prov_running.o \
	providers/libdefault.a providers/libcommon.a  -ldl -pthread
/opt/riscv-gnu/lib/gcc/riscv64-unknown-linux-gnu/13.2.0/../../../../riscv64-unknown-linux-gnu/bin/ld: /opt/riscv-gnu/lib/gcc/riscv64-unknown-linux-gnu/13.2.0/../../../../riscv64-unknown-linux-gnu/lib/libgcc_s.so.1: error adding symbols: file in wrong format

```

or also if I tried to add the required flags (-march=rv32imac -mabi=ilp32) I would get
```
make[1]: *** [Makefile:24431: test/cmp_asn_test] Error 1
/opt/riscv-gnu/lib/gcc/riscv64-unknown-linux-gnu/13.2.0/../../../../riscv64-unknown-linux-gnu/bin/ld: test/cipher_overhead_test(.text): relocation "test_vprintf_taperr+0x0 (type R_RISCV_CALL_PLT)" goes out of range
/opt/riscv-gnu/lib/gcc/riscv64-unknown-linux-gnu/13.2.0/../../../../riscv64-unknown-linux-gnu/bin/ld: test/libtestutil.a(libtestutil-lib-output.o): file class ELFCLASS64 incompatible with ELFCLASS32
/opt/riscv-gnu/lib/gcc/riscv64-unknown-linux-gnu/13.2.0/../../../../riscv64-unknown-linux-gnu/bin/ld: final link failed: file in wrong format
```

The toolchain am using supports building for both 32 and 64 bits libraries.
Therefore I would appreciate any help in this topic if anyone has had any luck compiling OpenSSL to rv32imac before, thanks beforehand. 

Notice: Am using **openSSL 3.0** because the newest release (3.2) presents with some problem during basic compilation such as ` error: 'rv64i_zkne_encrypt' undeclared (first use in this function)` but that is beside the point

**Comment by CharlesAverill** [comment]
Did you ever find a robust way to compile openssl for RISC-V? I'm having similar issues.
